agent/proxycfg/manager.go

```diff
@@ -0,0 +1,296 @@
+	state, ok := m.proxies[ns.ID]
+
+	if ok {
+		if !state.Changed(ns, token) {
+			// No change
+			return nil
+		}
+
+		// We are updating the proxy, close it's old state
+		state.Close()
+	}
+
+	var err error
+	state, err = newState(ns, token)
+	if err != nil {
+		return err
+	}
+
+	// Set the necessary dependencies
+	state.logger = m.cfg.Logger
+	state.cache = m.cfg.Cache
+	state.source = m.cfg.Source
+
+	ch, err := state.Watch()
+	if err != nil {
+		return err
+	}
+	m.proxies[ns.ID] = state
+
+	// Start a goroutine that will wait for changes and broadcast them to watchers.
+	go func(ch <-chan ConfigSnapshot) {
+		// Run until ch is closed
+		for snap := range ch {
+			m.notify(&snap)
+		}
+	}(ch)
+
+	return nil
+}
+
+// removeProxyService is called when a service deregisteres and frees all
+// resources for that service.
+func (m *Manager) removeProxyServiceLocked(proxyID string) {
+	state, ok := m.proxies[proxyID]
+	if !ok {
+		return
+	}
+
+	// Closing state will let the goroutine we started in Ensure finish since
+	// watch chan is closed.
+	state.Close()
+	delete(m.proxies, proxyID)
+
+	// We intentionally leave potential watchers hanging here - there is no new
+	// config for them and closing their channels might be indistinguishable from
+	// an error that they should retry. We rely for them to eventually give up
+	// (because they are in fact not running any more) and so the watches be
+	// cleaned up naturally.
+}
+
+func (m *Manager) notify(snap *ConfigSnapshot) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	watchers, ok := m.watchers[snap.ProxyID]
+	if !ok {
+		return
+	}
+
+	for _, ch := range watchers {
+		// Attempt delivery but don't let slow consumers block us forever. They
+		// might miss updates but it's better than breaking everything.
+		//
+		// TODO(banks): should we close their chan here to force them to eventually
+		// notice they are too slow? Not sure if it really helps.
+		select {
+		case ch <- snap:
+		case <-time.After(100 * time.Millisecond):
+		}
+	}
+}
+
+// Watch registers a watch on a proxy. It might not exist yet in which case this
+// will not fail, but no updates will be delivered until the proxy is
+// registered. If there is already a valid snapshot in memory, it will be
+// delivered immediately.
+func (m *Manager) Watch(proxyID string) (<-chan *ConfigSnapshot, CancelFunc) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// This buffering is crucial otherwise we'd block immediately trying to
+	// deliver the current snapshot below if we already have one.
+	ch := make(chan *ConfigSnapshot, 1)
+	watchers, ok := m.watchers[proxyID]
+	if !ok {
+		watchers = make(map[uint64]chan *ConfigSnapshot)
+	}
+	idx := uint64(len(watchers))
+	watchers[idx] = ch
+	m.watchers[proxyID] = watchers
+
+	// Deliver the current snapshot immediately if there is one ready
+	if state, ok := m.proxies[proxyID]; ok {
+		if snap := state.CurrentSnapshot(); snap != nil {
+			ch <- snap
```

Comment:
Would it be worth noting that this is safe to do without a `select` guard here because you absolutely know at this point that the channel is buffered and empty. It might be too easy to have a future refactor overlook this and cause an inadvertent deadlock.

Reply:
Might also be worth introducing a `select { default` with a ERROR log statement.